### DISCUSSION TITLE
Additional features

### DISCUSSION THREAD

**Discussion by DanielCarter1615** [ian-travers/nfsu-server#4] [Ideas]
Hey @ian-travers,

I saw in the old commits that there were some feature available in the past, which doesn't seem to work now anymore:

- 1. "Head to Head Send Challenge" bug. Create a 4 player game instead -> With that at least the game would not longer crash when trying to send a head to head challenge.
- 2.  Have the ability to change condition to join a specific room, e.g. a room where no uniques and no nos is allowed etc.

**Comment by ian-travers** [comment]
> "Head to Head Send Challenge" bug. Create a 4 player game instead

There was a little misunderstanding. "Create a 4 player game" was not a server feature, it was an appeal to the players.

Yes, the server has a feature for analyzing the level of the player's car. And in the code, a hard check was once set to enter a certain room. There are not so few variations of the car level, so the implementation of this feature in the `nfsu.conf`seems confusing.

**Comment by DanielCarter1615** [comment]
> There was a little misunderstanding. "Create a 4 player game" was not a server feature, it was an appeal to the players.

Okay, I misunderstood.

> Yes, the server has a feature for analyzing the level of the player's car. And in the code, a hard check was once set to enter a certain room. There are not so few variations of the car level, so the implementation of this feature in the nfsu.conf seems confusing.

I understand that it might be confusuing to set it in nfsu.conf. But can you explain me how I could hardcode some checks like that for my server, e.g. no nos, or no uniques?

**Comment by ian-travers** [comment]
@DanielCarter1615 You can uncomment lines [from 832](https://github.com/ian-travers/nfsu-server/blob/9bc06c9a5e31f1692c2389fc42a22f87f00dd891/NFSUServer2/nfsuserver.cpp#L832) to 856 and add below:
```
sprinf(log, "Car info: Brand = %u, WEI = %u, SUS = %u, ENG = %u, TUR = %u, NOS = %u, ECU = %u, TRA = %u, TIR = %u, BRA = %u", 
  Car_brand, Wei, Sus, Eng, Tur, Nos, Ecu, Tra, Tir, Bra);
Log(log);
```

Start the server and just enter any room. Open the server log. You should see a line like this:
```
[ Mon Aug 22 09:40:02 2022 ] Car info: Brand = 14, WEI = 7, SUS = 6 ENG = 7, TUR = 7, NOS = 6 ECU = 6, TRA = 6, TIR = 6, BRA = 6 
```
Let's see what that means.

**Comment by ian-travers** [comment]
In addition to the brand, we have nine performance parameters. Each is represented by three bits. The first two show the performance level. The third one points to the TJ's part.
Performance level:
```
00 - stock
01 - street
10 - pro
11 - extreme
```
TJ's part 
```
0 - not installed
1 - installed
```
So, the car above has all extreme performance levels and three TJ's parts: Weight reduction, Engine, Turbo. 6 = 110(bin), 7 = 111(bin)

**Comment by ian-travers** [comment]
You might set a bunch of checks. I think the best way to create a car helper. Provide functions in it.
```
bool hasTJPart(unsigned char part)
unsugned char partLevel(unsigned char part)

```
Then, when the player enters the certain room, check against his decoded car. 

```
if (partLevel(Eng) > 1 || partLevel(Tra) > 1 ...) {
  // kick out the room
}
```
or for TJ's parts

```
if (hasTJPart(Eng)  ||  hasTJPartl(Tra) ...) {
  // kick out the room
}
```

To kick player out see the commented [code](https://github.com/ian-travers/nfsu-server/blob/9bc06c9a5e31f1692c2389fc42a22f87f00dd891/NFSUServer2/nfsuserver.cpp#L862)

**Comment by DanielCarter1615** [comment]
Will check it out, thanks.

**Comment by DanielCarter1615** [comment]
I think I made it work. Not the cleanest, but it works: https://github.com/DanielCarter1615/nfsu-server/commit/bf7cce6fc9beced5af595510a3a239d7eb57c2c0

**Reply by ian-travers** [reply]
Cool. 
And I wouldn't worry about the cleanest code. The whole project is sloppy, repetitive and needs a good cleanup. This often happens when you want to go faster. Like mine in 2013.

**Reply by casasfernando** [reply]
Nice work @DanielCarter1615 . If I understand your code correctly, you are using fixed room names to set the conditions for that room (e.g. no nos)?

**Reply by DanielCarter1615** [reply]
@casasfernando Yes, exactly
